internal/pkg/state/signer.go

```diff
@@ -0,0 +1,61 @@
+package state
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/crypto"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/wallet"
+)
+
+type chainHeadTracker interface {
+	GetHead() block.TipSetKey
```

Comment:
This is the crux of the problem. Implicit head state access plus asynchrony can cause inconsistent state to be inspected by different bits of the code at different times. We're aiming to move all state access to use explicit keys. That means that this resolution cannot be implicit down deep in the system, but must be at time of use. Something like the SignatureValidator.

I would suggest leaving the old types.Signer interface alone and introducing a new one, with `Context`.

Reply:
I agree that implicit head state access is bad in general, but account actor signer addresses are immutable. Even if they weren't the question of what chain epoch am I signing this data at is somewhat non-nonsensical. I strongly doubt, if we were to ensure sub-components could provide a tip set key for signing, that they would ever use anything other than head, and that that key would every be used for another chain state operation.

So using the head tipset key retrieved only for this operation is a bad practice that we don't want to propagate, but in this case it has no significant consequences and helps eliminate a source of rather fiddly errors and overall complexity of the app. I didn't make this trade-off lightly.

Reply:
> the question of what chain epoch am I signing this data at

This has clear answers sometimes, like when signing a block header.